File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.java
class Solution {
    
    public boolean isValidPartition(int[] nums,int threshold,int divisor)
    {
        int sum=0;
        for(int i=0;i<nums.length;i++)
        {
            sum+=Math.ceil((double)nums[i]/divisor);
            if(sum>threshold)
            return false;
        }
        return true;
    }

    public int smallestDivisor(int[] nums, int threshold) 
    {
        int end=Arrays.stream(nums).max().orElse(0);;
        int start=1;

        int requiredDivisor=-1;
        while(start<=end)
        {
            int mid=start+(end-start)/2;
            if(isValidPartition(nums,threshold,mid))
            {
                requiredDivisor=mid;
                end=mid-1;
            }
            else
                start=mid+1;
        }  
        return requiredDivisor;
    }
}

File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int sumOfDivisors(vector<int>&nums,int mid){
        int n=nums.size();
        int sum=0;
        for(int i=0;i<n;i++){
            // sum=sum+ceil((double)nums[i]/mid);
            sum=sum+(nums[i]+mid-1)/mid;
        }
        return sum;
    }
    int smallestDivisor(vector<int>& nums, int threshold) {
        ios_base::sync_with_stdio(false);

        int maxEle=*max_element(nums.begin(),nums.end());

        int low=1,high=maxEle;
        int mid;
        while(low<=high){
            mid=low+(high-low)/2;
            int sum=sumOfDivisors(nums,mid);
            if(sum<=threshold){
                high=mid-1;
            }else{
                low=mid+1;
            }

        }
        return low;
    }
};